src/main/java/org/apache/datasketches/req/FloatBuffer.java

```diff
@@ -0,0 +1,418 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.datasketches.req;
+
+import static org.apache.datasketches.Criteria.GE;
+import static org.apache.datasketches.Criteria.GT;
+import static org.apache.datasketches.Criteria.LE;
+import static org.apache.datasketches.Criteria.LT;
+import static org.apache.datasketches.req.ReqSketch.LS;
+
+import java.util.Arrays;
+
+import org.apache.datasketches.BinarySearch;
+import org.apache.datasketches.Criteria;
+import org.apache.datasketches.SketchesArgumentException;
+
+/**
+ * A special buffer of floats specifically designed to support the ReqCompactor class.
+ *
+ * @author Lee Rhodes
+ */
+class FloatBuffer {
+  private float[] arr_;
+  private int count_;
+  private int capacity_;
+  private int delta_;
+  private boolean sorted_;
+  private boolean spaceAtBottom_;
+
+  /**
+   * Constructs an empty FloatBuffer with an initial capacity specified by
+   * the <code>capacity</code> argument.
+   *
+   * @param capacity the initial capacity.
+   * @param delta add space in increments of this size
+   * @param spaceAtBottom if true, create any extra space at the bottom of the buffer,
+   * otherwise, create any extra space at the top of the buffer.
+   */
+  FloatBuffer(final int capacity, final int delta, final boolean spaceAtBottom) {
+    arr_ = new float[capacity];
+    count_ = 0;
+    capacity_ = capacity;
+    delta_ = delta;
+    sorted_ = true;
+    spaceAtBottom_ = spaceAtBottom;
+  }
+
+  /**
+   * Copy Constructor
+   * @param buf the FloatBuffer to be copied into this one
+   */
+  FloatBuffer(final FloatBuffer buf) {
+    arr_ = buf.arr_.clone();
+    count_ = buf.count_;
+    capacity_ = buf.capacity_;
+    delta_ = buf.delta_;
+    sorted_ = buf.sorted_;
+    spaceAtBottom_ = buf.spaceAtBottom_;
+  }
+
+  /**
+   * Private construction from elements
+   * @param arr the array to be used directly as the internal array
+   * @param count the number of active elements in the given array
+   * @param delta add space in increments of this size
+   * @param capacity the initial capacity
+   * @param sorted true if already sorted
+   * @param spaceAtBottom if true, create any extra space at the bottom of the buffer,
+   * otherwise, create any extra space at the top of the buffer.
+   */
+  private FloatBuffer(final float[] arr, final int count, final int delta, final int capacity,
+      final boolean sorted, final boolean spaceAtBottom) {
+    arr_ = arr;
+    count_ = count;
+    capacity_ = capacity;
+    delta_ = delta;
+    sorted_ = sorted;
+    spaceAtBottom_ = spaceAtBottom;
+  }
```

Comment:
This is a work-around due to requirements of the REQ algorithm and the lack of a simple capability in Java-8: an **_efficient_**, reverse sort of primitive arrays.  If Java had such a capability, this hack would not have been necessary.  

Imagine you have a sorted array of 1K valid values (valid can include zero). To allow adding new values to the array we normally allocate space at the top of the array.  We keep two parameters, the capacity of the array and the number of currently valid values (for this I use "length" to correspond to "length" of Python vectors).  In the LRA mode of the sketch we will need to delete consecutive values at the top of the array and continue adding new valid values.  No problem. The extra space is always at the top.  Suppose we delete 10 from the top, the remaining 990 values remain in place.

However, when the sketch is in the HRA mode we need to delete values at the bottom of the array.  Now we have space at both the bottom and the top of the array. Where to add new values?  For space efficiency we always want these two spaces contiguous, and we would like to avoid having to move 990 values down 10 places.  So for this case it is easier to fill the array starting from the top and have the extra space for new values at the bottom.  Thus the name of the boolean.

Reply:
Your comment provides a clear explanation to my question, and has improved my understanding of why it is necessary.  I cannot think of a more efficient way to solve this, and agree that this is simplest.  Thanks!